event-buffer/event-buffer-core/src/main/java/uk/gov/justice/services/event/buffer/core/repository/service/ConsecutiveEventBufferService.java

```diff
@@ -0,0 +1,135 @@
+
+        final Optional<StreamStatus> currentStatus = streamStatusRepository.findByStreamId(streamId);
+        if (currentStatus.isPresent()) {
+            final Long currentVersion = currentStatus.get().getVersion();
+            if (incomingEventObsolete(incomingEventVersion, currentVersion)) {
+                LOGGER.trace(format("Message : %s is an obsolete version", toEnvelopeTraceString(incomingEvent)));
+                return Stream.empty();
+
+            } else if (incomingEventNotInOrder(incomingEventVersion, currentVersion)) {
+                LOGGER.trace(format("Message : %s is not consecutive, adding to buffer", toEnvelopeTraceString(incomingEvent)));
+                return addToBufferAndReturnEmptyStream(incomingEvent, streamId, incomingEventVersion);
+
+            } else {
+                LOGGER.trace(format("Message : %s version is valid sending stream to dispatcher", toEnvelopeTraceString(incomingEvent)));
+                return streamOfIncomingAndBufferedEvents(incomingEvent, streamId, incomingEventVersion);
+
+            }
+        } else {
+            if(incomingEventVersion == 1) {
+                LOGGER.trace(format("Message : %s is a new streamId registering with stream_status repo and sending to dispatcher", toEnvelopeTraceString(incomingEvent)));
+                streamStatusRepository.insert(new StreamStatus(streamId, incomingEventVersion));
+                return Stream.of(incomingEvent);
+            } else {
+                LOGGER.trace(format("Message : %s is a new streamId but version is not one so adding to buffer", toEnvelopeTraceString(incomingEvent)));
+                return addToBufferAndReturnEmptyStream(incomingEvent, streamId, incomingEventVersion);
+            }
+        }
+
+    }
+
+    private Stream<JsonEnvelope> streamOfIncomingAndBufferedEvents(JsonEnvelope incomingEvent, UUID streamId, Long incomingEventVersion) {
+        streamStatusRepository.update(new StreamStatus(streamId, incomingEventVersion));
+        return bufferedEvents(Stream.of(incomingEvent), streamBufferRepository.streamById(streamId), incomingEventVersion);
+    }
+
+    private Stream<JsonEnvelope> addToBufferAndReturnEmptyStream(JsonEnvelope incomingEvent, UUID streamId, Long incomingEventVersion) {
+        streamBufferRepository.insert(
+                new StreamBufferEvent(streamId,
+                        incomingEventVersion,
+                        jsonObjectEnvelopeConverter.asJsonString(incomingEvent)));
+
+        return Stream.empty();
+    }
+
+    private Stream<JsonEnvelope> bufferedEvents(Stream<JsonEnvelope> incomingMessage, final Stream<StreamBufferEvent> messageBuffer, final long currentVersion) {
+
+        System.out.println("Entered bufferedEvents()");
+
+        return Stream.concat(incomingMessage, consecutiveEventStreamFromBuffer(messageBuffer, currentVersion)
+                .peek(streamBufferEvent -> streamBufferRepository.remove(streamBufferEvent))
```

Comment:
Technically we're not modifying the stream here. We just peak into it to use it's element to call an external method.